Per-project option for soft clipping the output

purelygrey opened the pull request:
Implements #106. Project screen now has a "soft clip" option, which is saved per-project. I implemented [cubic soft clip](https://wiki.analog.com/resources/tools-software/sigmastudio/toolbox/nonlinearprocessors/standardcubic) to the best of my ability. It has four "modes" of varying hardness, which correspond to ceilings of roughly -1.5/-3/-6/-9db respectively. I also tried to gain compensate it, so switching it on increases overall loudness by the same amount. It goes before already existing hard clip, so turning it off (default) should sound exactly the same as before.

![softclip](https://github.com/user-attachments/assets/3df15b1e-9ef8-490e-9adb-869c6b02ab5e)

Videos of it in action: 
https://www.youtube.com/watch?v=W3pbEFrTkyg
https://www.youtube.com/watch?v=ZRu2YkMA9MM

All of the modes on sine wave at 0db peak: 
![image](https://github.com/user-attachments/assets/d37772b2-0b78-4e34-acc7-2cf8d3c25882)

Should cover most needs, I think. One existing quirk is the Master Volume being Input Gain of the clipper, it might be a good idea to add an "Output" volume that only goes downwards so one could balance their Piggy tracks volume-wise when preparing for dj sets, etc.

This is the first time I dabble in both C++ and audio processing, so this is ugly and I apologize. It works surprisingly well though. One ugly thing that remains is passing project's softclip variable into AudioOutDriver; at this point I don't know how to do it better. I had to set up `SetSoftclip()` method in both Mixer and AudioOut to pass the variable, and I had to create an empty `DummyAudioOut::SetSoftclip` method to circumvent the "invalid new expression of abstract class type <...> because the following virtual functions are pure within" compilation error.

Tested it under Linux x64, on Powkiddy V90 (bittboy), and Miyoo Mini (miyoo).

Let me know what you think! I appreciate any input or insight on how to make things better :)

I also fixed spacebar mapping for X64.

djdiskmachine commented:
Exciting!
Sounds really good, I'll have a look at it more later today =)

djdiskmachine commented:
I've added a pre-commit hook you can add to your own repo.
install it with cp pre-commit .git/hooks/ after rebasing on master =)

djdiskmachine commented on sources/Services/Audio/AudioOutDriver.cpp:
```diff
@@ -80,34 +90,91 @@ void AudioOutDriver::clipToMix() {
+    float x;
+    float alpha=1;
+    float twoThirds=2.0/3.0;
+    float gainCompensation=1.0/(alpha*twoThirds)-0.05;
```
Would be _really_ sweet to have an adjustable gain here =3

djdiskmachine commented on sources/Services/Audio/AudioOutDriver.cpp:
```diff
@@ -80,34 +90,91 @@ void AudioOutDriver::clipToMix() {
+		fixed fl_32767=fp2fl(f_32767);
+		fixed fl_m32768=fp2fl(f_m32768);
+
+    float x;
```
I've been going through this section a few times now. I guess it's a standard cubic symmetric clipping algorithm? In that case, could this variable have a more descriptive name such as clippedVal? =)

purelygrey replied:
Tried it, but I think it gets confusing, since it's not really clipped half of the time. Added a link to the algorithm description instead, it does make sense in context

djdiskmachine commented on sources/Services/Audio/AudioOutDriver.cpp:
```diff
@@ -66,6 +68,14 @@ void AudioOutDriver::prepareMixBuffers() {
 } ;
 
+void AudioOutDriver::SetSoftclip(int clip) {
+  if (clip == 1) {
```
softclip_ = (clip == 1) ? true : false;

djdiskmachine reviewed (commented):
I like it a lot!
Needs a little bit of work with readability, variable names in the clipping part is a little bit confusing. Other than that it's a great addition!

djdiskmachine commented on sources/Services/Audio/AudioOutDriver.cpp:
```diff
@@ -80,34 +90,91 @@ void AudioOutDriver::clipToMix() {
-			s2+=offset;
-		} ;
+
+      if (softclip_ == 0) {
```
Actually I'd prefer if it was if (softClip_) and the conditions were reversed.

purelygrey commented on sources/Application/Mixer/MixerService.cpp:
```diff
@@ -161,8 +161,6 @@ bool MixerService::Clipped() {
 } ;
 
 void MixerService::SetMasterVolume(int vol) {
-  Mixer *mixer=Mixer::GetInstance();
```
It didn't appear to be used in this method so I got rid of it, probably shouldn't have honestly

purelygrey replied:
Restored it

purelygrey commented:
@djdiskmachine I took the advice to use the git hook literally and rebased my commits and ended up force pushing into my fork, you'll have to hard reset your local branch, sorry if you didn't mean it!

Improved most things:

- Instead of an on/off switch there are 4 "modes" now, with different hardness levels. They boost the signal by ~1.5/3/6/9db. Since Master volume is the input gain and the output gets boosted to 0db after softclipping, I decided not to do a separate input gain/output gain for now, I like the simplicity of one gain control. Thought about doing a 0-100 integer field, but flavour words seemed way cooler and a few usable values instead of a vague number seem better too
- Reorganized clipping logic, now hard clip always goes after soft clip, just in case there are overshoots after gain compensation and converting to fixed point (honestly don't know if that might happen or not, mostly guessing)
- Cleaned things up, should be a lot more readable now, especially the clipping part, should address most if not all problems you mentioned (thanks!)
- Fixed gain compensation math and polarity problems
- Brought back the reference to mixer in MixerService

I also tested bittboy & miyoo mini builds of the previous commit and they worked fine

Let me know what you think! Here's a 0db sine wave in all of the modes: 
![softcliptypes](https://github.com/user-attachments/assets/cdd4515e-e198-424b-9177-8c770f3a3529)

djdiskmachine commented:
Great work! Can't wait to try this out. :)
Not sure why clang is failing again, maybe you didn't get the rules I added yesterday

djdiskmachine commented:
Ok now if you apply the git diff at the bottom here it should be all good 
https://github.com/djdiskmachine/LittleGPTracker/actions/runs/11173977362/job/31062911515?pr=145

djdiskmachine commented:
I guess I need to fix my git hook, sorry about that.

purelygrey commented:
> I guess I need to fix my git hook, sorry about that.

No problem, I had to apply two diffs, seems to be fine now
